mlflow/models/evaluation/default_evaluator.py

```diff
@@ -614,21 +637,116 @@ def plot_pr_curve():
 
         self._log_pandas_df_artifact(per_class_metrics_collection_df, "per_class_metrics")
 
-    def _evaluate_custom_metrics(self):
+    def _log_custom_metric_artifact(self, artifact_name, raw_artifact, custom_metric_tuple):
+        """
+        This function logs and returns a custom metric artifact. Two cases:
+            - The provided artifact is a path to a file, the function will make a copy of it with
+              a formatted name in a temporary directory and call mlflow.log_artifact.
+            - Otherwise: will attempt to save the artifact to an temporary path with an inferred
+              type. Then call mlflow.log_artifact.
+
+        :param artifact_name: the name of the artifact
+        :param raw_artifact:  the object representing the artifact
+        :param custom_metric_tuple: an instance of the _CustomMetric namedtuple
+        :return: EvaluationArtifact
+        """
+
+        exception_and_warning_header = (
+            f"Custom metric function '{custom_metric_tuple.name}' at index "
+            f"{custom_metric_tuple.index} in the `custom_metrics` parameter"
+        )
+
+        inferred_from_path, inferred_type, inferred_ext = _infer_artifact_type_and_ext(
+            artifact_name, raw_artifact, custom_metric_tuple
+        )
+        artifact_file_name = _gen_log_key(artifact_name, self.dataset_name) + inferred_ext
+        artifact_file_local_path = self.temp_dir.path(artifact_file_name)
```

Comment:
What happens if there's a collision on `artifact_file_local_path`? i.e. 2 or more custom metric functions are creating the same artifact?

Reply:
Right, I think it's a good idea to throw an exception to let the user know if there's a collision.